system/jlib/jfile.cpp

```diff
@@ -7106,3 +7111,244 @@ const FileSystemProperties & queryFileSystemProperties(const char * filename)
     else
         return linuxFileSystemProperties;
 }
+
+
+#if defined(__linux__)
+class CFileEventWatcher : public CInterfaceOf<IFileEventWatcher>, implements IThreaded
+{
+    class CMonitoredItem : public CInterface
+    {
+    public:
+        CMonitoredItem(const char *_filename, int _watchFd, FileWatchEvents _events) : filename(_filename), watchFd(_watchFd), events(_events)
+        {
+        }
+        std::string filename;
+        int watchFd = -1;
+        FileWatchEvents events = FileWatchEvents::none;
+        bool pendingRemoval = false;
+    };
+    int inotifyFd = -1;
+    int pipefd[2] = {-1, -1};
+    pollfd toPoll[2];
+    std::unordered_map<int, Linked<CMonitoredItem>> monitoredFilesByFd;
+    std::unordered_map<std::string, Linked<CMonitoredItem>> monitoredFiles;
+    std::unordered_set<int> pendingRemovals;
+    CriticalSection crit;
+    FileWatchFunc callback;
+    CThreaded threaded;
+    bool stopped = true;
+
+    void handle_events(int fd)
+    {
+        byte buf[4096] __attribute__ ((aligned(__alignof__(struct inotify_event))));
+
+        while (true)
+        {
+            ssize_t len = read(fd, buf, sizeof(buf));
+            if (len == -1 && errno != EAGAIN)
+                throw makeErrnoException("CFileEventWatcher::handle_events - read");
+
+            if (len <= 0)
+                break;
+
+            const byte *ptr = buf;
+
+            do
+            {
+                const struct inotify_event *event = (const struct inotify_event *) ptr;
+                std::string filename;
+                if (event->len) // if watching directory event->name is populated with file within that dir.
+                    filename = event->name;
+                else
+                {
+                    // if watching a file, lookup filename from event handle
+                    CriticalBlock b(crit);
+                    auto iter = monitoredFilesByFd.find(event->wd);
+                    if (iter != monitoredFilesByFd.end())
+                    {
+                        filename = iter->second->filename;
+                        break;
+                    }
+                }
+                /* NB: in this impl. inotify_event flags are same as FileWatchEvents bit masks
+                 * If other implementations differ, they will need mapping to FileWatchEvents
+                 */
+                FileWatchEvents events = static_cast<FileWatchEvents>(event->mask);
+
+                callback(filename.c_str(), events);
+
+                ptr += sizeof(struct inotify_event) + event->len;
+            }
+            while (ptr < (buf + len));
+        }
+    }
+    void checkClose(int h)
+    {
+        if (-1 != h)
+            close(h);
+    }
+public:
+    CFileEventWatcher(FileWatchFunc _callback) : callback(_callback), threaded("CFileEventWatcher", this)
+    {
+        inotifyFd = inotify_init1(IN_NONBLOCK);
```

Comment:
Should this also set IN_CLOEXEC?

Reply:
yes think it should, so should the pipe handles, and I see that can use pipe2 to create those without the fcntl code.
Will change.